modin/pandas/groupby.py

```diff
@@ -194,41 +194,48 @@ def ndim(self):
         return 2  # ndim is always 2 for DataFrames
 
     def shift(self, periods=1, freq=None, axis=0, fill_value=None):
-        def _shift(periods, freq, axis, fill_value, is_set_nan_rows=True):
+        def _shift(data, periods, freq, axis, fill_value, is_set_nan_rows=True):
             from .dataframe import DataFrame
 
-            result = self._df.shift(periods, freq, axis, fill_value)
+            result = data.shift(periods, freq, axis, fill_value)
 
             if (
                 is_set_nan_rows
                 and isinstance(self._by, BaseQueryCompiler)
                 and (
                     # Check using `issubset` is effective only in case of MultiIndex
-                    set(self._by.columns).issubset(list(self._df.columns))
+                    set(self._by.columns).issubset(list(data.columns))
                     if isinstance(self._by.columns, pandas.MultiIndex)
                     else len(
                         self._by.columns.unique()
                         .sort_values()
-                        .difference(self._df.columns.unique().sort_values())
+                        .difference(data.columns.unique().sort_values())
                     )
                     == 0
                 )
                 and DataFrame(query_compiler=self._by.isna()).any(axis=None)
             ):
-                mask_nan_rows = self._df[self._by.columns].isna().any(axis=1)
+                mask_nan_rows = data[self._by.columns].isna().any(axis=1)
                 # drop NaN groups
                 result = result.loc[~mask_nan_rows]
             return result
 
         if freq is None and axis == 1 and self._axis == 0:
-            result = _shift(periods, freq, axis, fill_value)
+            no_by_data = (
+                self._df.drop(columns=self._by.columns)
+                if isinstance(self._by, BaseQueryCompiler)
+                else self._df
+            )
```

Comment:
new pandas version does not allow to shift from/to "by" columns, so filtering the source frame from "by" columns before shifting
```python
>>> data = {
...     "col1": [0, 1, 2, 3],
...     "col2": [4, 5, np.NaN, 7],
...     "col3": [np.NaN, np.NaN, 12, 10],
...     "col4": [17, 13, 16, 15],
...     "col5": [-4, -5, -6, -7],
... }
>>> pandas125.DataFrame(data).groupby(["col1", "col4", "col5"]).shift(axis=1)
   col1  col2  col3  col4  col5
0   NaN     0   4.0   NaN    17
1   NaN     1   5.0   NaN    13
2   NaN     2   NaN  12.0    16
3   NaN     3   7.0  10.0    15
>>> pandas130.DataFrame(data).groupby(["col1", "col4", "col5"]).shift(axis=1)
   col2  col3
0   NaN   4.0
1   NaN   5.0
2   NaN   NaN
3   NaN   7.0
```

Reply:
has rewritten in a more general way: now it drops 'by' columns not only in the "`BaseQueryCompiler` case"